File: backend/app/models/maintenance.py
import uuid
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Enum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base
import enum


class MaintenanceCategory(str, enum.Enum):
    plumbing = "plumbing"
    electrical = "electrical"
    hvac = "hvac"
    appliance = "appliance"
    structural = "structural"
    other = "other"


class RequestPriority(str, enum.Enum):
    low = "low"
    medium = "medium"
    high = "high"
    urgent = "urgent"


class RequestStatus(str, enum.Enum):
    open = "open"
    in_progress = "in_progress"
    resolved = "resolved"
    closed = "closed"
    rejected = "rejected"


class StaffWorkStatus(str, enum.Enum):
    assigned = "assigned"
    in_progress = "in_progress"
    completed = "completed"


class MaintenanceRequest(Base):
    __tablename__ = "maintenance_requests"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    unit_id = Column(UUID(as_uuid=True), ForeignKey("units.id", ondelete="CASCADE"), nullable=False, index=True)
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    category = Column(Enum(MaintenanceCategory), nullable=False)
    title = Column(String(200), nullable=False)
    description = Column(Text, nullable=False)
    priority = Column(Enum(RequestPriority), nullable=False, default=RequestPriority.medium, index=True)
    status = Column(Enum(RequestStatus), nullable=False, default=RequestStatus.open, index=True)
    attachment_url = Column(Text, nullable=True)
    resolved_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    unit = relationship("Unit", back_populates="maintenance_requests")
    tenant = relationship("User", foreign_keys=[tenant_id])
    assignments = relationship("MaintenanceAssignment", back_populates="request", cascade="all, delete-orphan")


class MaintenanceAssignment(Base):
    __tablename__ = "maintenance_assignments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    request_id = Column(UUID(as_uuid=True), ForeignKey("maintenance_requests.id", ondelete="CASCADE"), nullable=False, index=True)
    staff_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    assigned_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    notes = Column(Text, nullable=True)
    status = Column(Enum(StaffWorkStatus), nullable=False, default=StaffWorkStatus.assigned)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    request = relationship("MaintenanceRequest", back_populates="assignments")
    staff = relationship("User", foreign_keys=[staff_id])
    assigner = relationship("User", foreign_keys=[assigned_by])

File: frontend/src/types/maintenance.ts
export type MaintenanceCategory =
  | 'plumbing'
  | 'electrical'
  | 'hvac'
  | 'appliance'
  | 'structural'
  | 'other';

export type RequestPriority = 'low' | 'medium' | 'high' | 'urgent';

export type RequestStatus =
  | 'open'
  | 'in_progress'
  | 'resolved'
  | 'closed'
  | 'rejected';

export type StaffWorkStatus = 'assigned' | 'in_progress' | 'completed';

export interface UserInfo {
  id: string;
  email: string;
  full_name: string | null;
}

export interface MaintenanceAssignment {
  id: string;
  request_id: string;
  staff_id: string;
  assigned_by: string;
  notes: string | null;
  status: StaffWorkStatus;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  staff: UserInfo | null;
  assigner: UserInfo | null;
}

export interface MaintenanceRequest {
  id: string;
  unit_id: string;
  tenant_id: string;
  category: MaintenanceCategory;
  title: string;
  description: string;
  priority: RequestPriority;
  status: RequestStatus;
  attachment_url: string | null;
  resolved_at: string | null;
  created_at: string;
  updated_at: string;
  tenant: UserInfo | null;
  assignments: MaintenanceAssignment[];
}

export interface MaintenanceRequestCreate {
  unit_id: string;
  category: MaintenanceCategory;
  title: string;
  description: string;
  priority?: RequestPriority;
  attachment_url?: string | null;
}

export interface MaintenanceRequestUpdate {
  status?: RequestStatus;
  priority?: RequestPriority;
  title?: string;
  description?: string;
}

export interface MaintenanceAssignmentCreate {
  staff_id: string;
  notes?: string | null;
}

export interface MaintenanceAssignmentUpdate {
  status?: StaffWorkStatus;
  notes?: string | null;
}

export interface StaffAssignment {
  id: string;
  request_id: string;
  staff_id: string;
  assigned_by: string;
  notes: string | null;
  status: StaffWorkStatus;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  request: MaintenanceRequest | null;
  assigner: UserInfo | null;
}